frontend/server/libs/dao/Problemsets.dao.php

```diff
@@ -39,30 +39,35 @@ public static function getProblemsetContainer($problemset_id) {
      *  Check if a submission is before the deadline.
      *  No one, including admins, can submit after the deadline.
      */
-    public static function isLateSubmission($container) {
+    public static function isLateSubmission(Object $container) : bool {
         return isset($container->finish_time) &&
                (Time::get() > strtotime($container->finish_time));
     }
 
-    public static function insideSubmissionWindow($container, $identity_id) {
+    public static function insideSubmissionWindow(
+        Object $container,
+        int $identityId,
+        ?ProblemsetIdentities $problemsetIdentities = null
+    ) : bool {
         if (isset($container->finish_time)) {
             if (Time::get() > strtotime($container->finish_time) ||
                 Time::get() < strtotime($container->start_time)) {
-                return false;
+                if (is_null($problemsetIdentities)) {
+                    return false;
+                }
+                return Time::get() <= strtotime($problemsetIdentities->end_time);
             }
         }
 
         if (!isset($container->window_length)) {
             return true;
         }
 
-        $problemset_identity = ProblemsetIdentitiesDAO::getByPK(
-            $identity_id,
-            $container->problemset_id
-        );
-        $first_access_time = $problemset_identity->access_time;
-
-        return Time::get() <= strtotime($first_access_time) + $container->window_length * 60;
+        if (is_null($problemsetIdentities)) {
+            return false;
+        }
+        return is_null($problemsetIdentities) ?:
```

Comment:
```suggestion
        return !is_null($problemsetIdentities) &&
```

así puedes borrar L66-L68.

Reply:
Listo